src/pybind/mgr/devicehealth/module.py

```diff
@@ -0,0 +1,249 @@
+    def serve(self):
+        self.log.info("Starting")
+        while self.run:
+            self.refresh_config()
+
+            # TODO normalize/align sleep interval
+            sleep_interval = int(self.scrape_frequency)
+
+            self.log.debug('Sleeping for %d seconds', sleep_interval)
+            ret = self.event.wait(sleep_interval)
+            self.event.clear()
+
+            # in case 'wait' was interrupted, it could mean config was changed
+            # by the user; go back and read config vars
+            if ret:
+                continue
+
+            self.log.debug('Waking up [%s]',
+                           "active" if self.active else "inactive")
+            if not self.active:
+                continue
+            self.log.debug('Running')
+
+    def open_connection(self):
+        pools = self.rados.list_pools()
+        is_pool = False
+        for pool in pools:
+            if pool == self.pool_name:
+                is_pool = True
+                break
+        if not is_pool:
+            self.rados.create_pool(self.pool_name)
+            self.log.debug('create %s pool' % self.pool_name)
+        ioctx = self.rados.open_ioctx(self.pool_name)
+        return (ioctx)
+
+    def scrape_osd(self, osd_id):
+        ioctx = self.open_connection()
+        raw_smart_data = self.do_scrape_osd(osd_id, ioctx)
+        if raw_smart_data:
+            for device, raw_data in raw_smart_data.items():
+                data = self.extract_smart_features(raw_data)
+                self.put_device_metrics(ioctx, device, data)
+        ioctx.close()
+        return (0, "", "")
+
+    def scrape_all(self):
+        osdmap = self.get("osd_map")
+        assert osdmap is not None
+        ioctx = self.open_connection()
+        did_device = {}
+        for osd in osdmap['osds']:
+            osd_id = osd['osd']
+            raw_smart_data = self.do_scrape_osd(osd_id, ioctx)
+            if not raw_smart_data:
+                continue
+            for device, raw_data in raw_smart_data.items():
+                if device in did_device:
+                    self.log.debug('skipping duplicate %s' % device)
+                    continue
+                did_device[device] = 1
+                data = self.extract_smart_features(raw_data)
+                self.put_device_metrics(ioctx, device, data)
+
+        ioctx.close()
+        return (0, "", "")
+
+    def scrape_device(self, devid):
+        r = self.get("device " + devid)
+        if not r or 'device' not in r.keys():
+            return (-errno.ENOENT, '', 'device ' + devid + ' not found')
+        daemons = r['device'].get('daemons', [])
+        osds = [int(r[4:]) for r in daemons if r.startswith('osd.')]
+        if not osds:
+            return (-errno.EAGAIN, '',
+                    'device ' + devid + ' not claimed by any active OSD daemons')
+        osd_id = osds[0]
+        ioctx = self.open_connection()
+        raw_smart_data = self.do_scrape_osd(osd_id, ioctx)
```

Comment:
I think devid is missing in this call, i.e.:
`raw_smart_data = self.do_scrape_osd(osd_id, ioctx, devid)`